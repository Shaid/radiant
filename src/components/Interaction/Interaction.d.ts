export interface IState {
  active: boolean
}

export interface IProps {
  actions: Array<IAction>
  children?: string[] | string | undefined
  description?: string
}

export interface IAction {
  label: string
  callback: Function
}

export interface IDefaultProps {
}

export type PropsWithDefaults = IProps & IDefaultProps
