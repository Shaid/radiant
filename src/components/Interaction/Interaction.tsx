import React from 'react'
import styled, { breakpoints, theme } from 'theme'
import { rgba } from 'polished'
import { defaultButtonStyle } from 'theme/helpers'

import { IProps, IDefaultProps, IState } from './Interaction.d'

interface IButtonStyled {
  active: boolean
}

interface IActionsStyled {
  active: boolean
  count: number
}

const Interaction = styled.div`
  position: relative;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
`

const Button = styled.button`
  position: relative;
  ${defaultButtonStyle}
  ${(props: IButtonStyled) => (props.active ? `z-index: 500; box-shadow: inset 0 0 0 0 ${theme.primaryBackground}, 0 15vh 25vh 25vh ${theme.primaryBackground}; border-radius: 5vh;` : '')}
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
`

const positionItems = (count: number, active: boolean) => {
  let positions = ''

  for (let i = 1; i <= count; i++) {
    positions += `
    &:nth-child(${i}) {
      transform: translateY(${active ? 2.0 * i : 0}rem);

      @media ${breakpoints.desktop} {
        transform: translateY(${active ? 1.25 * i : 0}rem);
      }
    }
    `
  }

  return `
  > * {
    display: block;
    position: absolute;
    font-size: 0.9rem;
    min-width: 100px;
    max-width: 100vw;
    text-decoration: none;
    transition: transform 0.2s ease-in;
    left: 0;
    top: 10px;
    ${positions}

  }

  > button {
    ${defaultButtonStyle}
    cursor: ${active ? 'pointer' : 'default'};
    font-size: 1.3rem;
    font-variant: small-caps;

    :hover {
      background-color: ${rgba(theme.primaryColourDecorate, 0.2)};
      border-radius: 5px;
    }

    @media ${breakpoints.desktop} {
      font-size: 1.1rem;
    }
  }
  `
}

const ActiveView = styled.div`
  display: flex;
  flex-flow: column nowrap;
`

const Description = styled.div`
  font-style: italic;
  position: absolute;
  font-size: 0.9rem;
  text-decoration: none;
  transition: transform 0.2s ease-in;
  left: 0;
  top: 30px;
  width: 100vw;

  @media ${breakpoints.desktop} {
    font-size: 0.9rem;
  }
`

const ActionsList = styled.nav`
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  align-self: center;
  position: absolute;
  top: 0;
  width: 100px;
  opacity: ${(props: IActionsStyled) => (props.active ? '1' : '0')};
  visibility: ${(props: IActionsStyled) => (props.active ? 'visible' : 'hidden')};
  border-radius: 50%;
  transition: opacity 0.2s ease-in, visibility 0.2s ease-in, transform 0.2s ease-in;
  ${(props: IActionsStyled) => (props.active ? 'z-index: 500;' : 'z-index: -1;')}
  ${(props: IActionsStyled) => positionItems(props.count, props.active)}
`


export default class extends React.PureComponent<IProps, IState> {
  static defaultProps: Partial<IDefaultProps> = {}

  constructor(props: IProps, state: IState) {
    super(props, state)

    this.state = {
      active: false
    }
  }

  setInactive = (event: Event) => {
    event.stopImmediatePropagation()
    this.setActive(false)
  }

  setActive(active: boolean) {
    if (active) {
      document.addEventListener('click', this.setInactive, false)
    } else {
      document.removeEventListener('click', this.setInactive, false)
    }

    // add observer to check if offscreen.
    const ioOptions = {
      root: null,
      rootMargin: '10px',
      threshold: 1.0
    }
    const intersectionObserver = new IntersectionObserver((entries) => {
      entries.forEach(entry => {
        if (!entry.isIntersecting || entry.intersectionRatio < 1) {
          const el:any = entry.target
          // const elStyle = window.getComputedStyle(el!)
          const x = Math.min(0, entry.boundingClientRect.left, entry.boundingClientRect.right)
          const y = Math.min(0, entry.boundingClientRect.top, entry.boundingClientRect.bottom)

          if (x < 0) { console.log(`left ${-x}`); el.style.left = `${-x}px` }
          if (y < 0) { el.style.top = `${-y}px` }
        }
      })
    }, ioOptions)

    Array.from(this.actionsRef.current.children).forEach((child: Element) => {
      intersectionObserver.observe(child)
    })

    this.setState({
      active
    })
  }

  actionsRef:any = React.createRef()

  render() {
    const { actions, children, description } = this.props
    const { active } = this.state

    const actionsList = actions.map((action) => (
      <button
        onClick={(event) => {
          if (active) {
            event.stopPropagation()
            action.callback()
            this.setActive(false)
          }
        }}
        key={action.label}
        title={action.label}
      >
        {action.label}
      </button>
    ))

    return (
      <Interaction>
        <Button
          active={active}
          onClick={(event) => {
            event.stopPropagation()
            this.setActive(true)
          }}
        >
          {children}
        </Button>
        <ActiveView>
          <ActionsList innerRef={this.actionsRef} count={actionsList.length + 1} active={active} >
            <Description>
              {description}
            </Description>
            {actionsList}
          </ActionsList>
        </ActiveView>
      </Interaction>
    )
  }
}
