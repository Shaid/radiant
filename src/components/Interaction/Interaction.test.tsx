import React from 'react'
import { mount, shallow } from 'enzyme'
import 'intersection-observer'
import Interaction from '.'

describe('<Action />', () => {
  it('renders', () => {
    const wrapper = shallow(<Interaction actions={[]} />)
    expect(wrapper).toMatchSnapshot()
  })

  // skipped until I mock IntersectionObserver
  it.skip('triggers its action when clicked', () => {
    const click = jest.fn()

    const wrapper = mount(<Interaction actions={[]} />)
    wrapper.find('button').simulate('click')
    expect(click).toHaveBeenCalled()
  })

  it('can trigger a subaction when clicked, and then an action is clicked ', () => {
    const action = jest.fn()

    const actions = [
      { label: 'test action', callback: action }
    ]

    const wrapper = mount(<Interaction actions={actions}>Some thing!</Interaction>)
    wrapper.find('button').first().simulate('click')
    wrapper.find('[title="test action"]').simulate('click')
    expect(action).toHaveBeenCalled()
  })
})
