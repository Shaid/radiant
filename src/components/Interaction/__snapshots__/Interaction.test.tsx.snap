// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Action /> renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <_class
    actions={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef
          active={false}
          onClick={[Function]}
        />,
        <ForwardRef>
          <ForwardRef
            active={false}
            count={1}
          >
            <ForwardRef />
          </ForwardRef>
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
          "children": undefined,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": false,
            "rules": Array [
              "
  position: relative;
  ",
              "
color: #dadada;
margin: 0;
padding: 0;
background: transparent;
border: none;
font-family: 'EB Garamond', serif;
font-size: 1rem;
cursor: pointer;
text-decoration: underline solid rgba(90,90,90,0.75);
user-select: none;
-webkit-tap-highlight-color: transparent;

:focus {
  outline: none;
}
",
              "
  ",
              [Function],
              "
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
",
            ],
          },
          "displayName": "styled.button",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef
            active={false}
            count={1}
          >
            <ForwardRef />
          </ForwardRef>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": false,
            "children": Array [
              <ForwardRef />,
              Array [],
            ],
            "count": 1,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": true,
                  "rules": Array [
                    "
  font-style: italic;
  position: absolute;
  font-size: 0.9rem;
  text-decoration: none;
  transition: transform 0.2s ease-in;
  left: 0;
  top: 30px;
  width: 100vw;

  @media ",
                    "all and (orientation: landscape)",
                    " {
    font-size: 0.9rem;
  }
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": false,
              "rules": Array [
                "
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  align-self: center;
  position: absolute;
  top: 0;
  width: 100px;
  opacity: ",
                [Function],
                ";
  visibility: ",
                [Function],
                ";
  border-radius: 50%;
  transition: opacity 0.2s ease-in, visibility 0.2s ease-in, transform 0.2s ease-in;
  ",
                [Function],
                "
  ",
                [Function],
                "
",
              ],
            },
            "displayName": "styled.nav",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "nav",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "rules": Array [
              "
  display: flex;
  flex-flow: column nowrap;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
  position: relative;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            active={false}
            onClick={[Function]}
          />,
          <ForwardRef>
            <ForwardRef
              active={false}
              count={1}
            >
              <ForwardRef />
            </ForwardRef>
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": false,
            "children": undefined,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": false,
              "rules": Array [
                "
  position: relative;
  ",
                "
color: #dadada;
margin: 0;
padding: 0;
background: transparent;
border: none;
font-family: 'EB Garamond', serif;
font-size: 1rem;
cursor: pointer;
text-decoration: underline solid rgba(90,90,90,0.75);
user-select: none;
-webkit-tap-highlight-color: transparent;

:focus {
  outline: none;
}
",
                "
  ",
                [Function],
                "
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
",
              ],
            },
            "displayName": "styled.button",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "button",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef
              active={false}
              count={1}
            >
              <ForwardRef />
            </ForwardRef>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "active": false,
              "children": Array [
                <ForwardRef />,
                Array [],
              ],
              "count": 1,
            },
            "ref": Object {
              "current": null,
            },
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bxivhb",
                    "isStatic": true,
                    "rules": Array [
                      "
  font-style: italic;
  position: absolute;
  font-size: 0.9rem;
  text-decoration: none;
  transition: transform 0.2s ease-in;
  left: 0;
  top: 30px;
  width: 100vw;

  @media ",
                      "all and (orientation: landscape)",
                      " {
    font-size: 0.9rem;
  }
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bxivhb",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-ifAKCX",
                "isStatic": false,
                "rules": Array [
                  "
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  align-self: center;
  position: absolute;
  top: 0;
  width: 100px;
  opacity: ",
                  [Function],
                  ";
  visibility: ",
                  [Function],
                  ";
  border-radius: 50%;
  transition: opacity 0.2s ease-in, visibility 0.2s ease-in, transform 0.2s ease-in;
  ",
                  [Function],
                  "
  ",
                  [Function],
                  "
",
                ],
              },
              "displayName": "styled.nav",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-ifAKCX",
              "target": "nav",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  flex-flow: column nowrap;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "rules": Array [
            "
  position: relative;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
